wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/monitoring/MonitoringRdsHostListProvider.java

```diff
@@ -0,0 +1,172 @@
+    this.writerTopologyQuery = writerTopologyQuery;
+    this.highRefreshRateNano = TimeUnit.MILLISECONDS.toNanos(
+        CLUSTER_TOPOLOGY_HIGH_REFRESH_RATE_MS.getLong(this.properties));
+  }
+
+  /**
+   * Clear topology cache for the current cluster.
+   */
+  public void clear() {
+    topologyCache.remove(this.clusterId);
+  }
+
+  public static void clearCache() {
+    monitors.getEntries().entrySet().forEach(entry -> {
+      try {
+        entry.getValue().close();
+      } catch (Exception ex) {
+        // ignore
+      }
+    });
+    monitors.clear();
+    topologyCache.clear();
+    primaryClusterIdCache.clear();
+    suggestedPrimaryClusterIdCache.clear();
+  }
+
+  @Override
+  protected void init() throws SQLException {
+    super.init();
+    this.initMonitor();
+  }
+
+  protected ClusterTopologyMonitor initMonitor() {
+    return monitors.computeIfAbsent(this.clusterId,
+        (key) -> new ClusterTopologyMonitorImpl(
+            key, topologyCache, this.initialHostSpec, this.properties, this.pluginService,
+            this.hostListProviderService, this.clusterInstanceTemplate,
+            this.refreshRateNano, this.highRefreshRateNano, TOPOLOGY_CACHE_EXPIRATION_NANO,
+            this.topologyQuery,
+            this.writerTopologyQuery,
+            this.nodeIdQuery),
+        MONITOR_EXPIRATION_NANO);
+  }
+
+  @Override
+  protected List<HostSpec> queryForTopology(final Connection conn) throws SQLException {
+    ClusterTopologyMonitor monitor = monitors.get(this.clusterId, MONITOR_EXPIRATION_NANO);
+    if (monitor == null) {
+      monitor = this.initMonitor();
+    }
+    try {
+      return monitor.forceRefresh(conn, defaultTopologyQueryTimeoutMs);
+    } catch (TimeoutException ex) {
+      return null;
+    }
+  }
+
+  @Override
+  protected void clusterIdChanged(final String oldClusterId) {
+    final ClusterTopologyMonitor existingMonitor = monitors.get(this.clusterId, MONITOR_EXPIRATION_NANO);
+    if (existingMonitor != null) {
+      monitors.computeIfAbsent(this.clusterId, (key) -> existingMonitor, MONITOR_EXPIRATION_NANO);
+      assert monitors.get(this.clusterId, MONITOR_EXPIRATION_NANO) == existingMonitor;
+      existingMonitor.setClusterId(this.clusterId);
+      monitors.remove(oldClusterId);
+    }
+
+    final List<HostSpec> existingHosts = topologyCache.get(oldClusterId);
+    if (existingHosts != null) {
+      topologyCache.put(this.clusterId, existingHosts, TOPOLOGY_CACHE_EXPIRATION_NANO);
+    }
+  }
+
+  @Override
+  public List<HostSpec> forceRefresh(final boolean writerImportant, final long timeoutMs)
+      throws SQLException, TimeoutException {
+
+    ClusterTopologyMonitor monitor = monitors.get(this.clusterId, MONITOR_EXPIRATION_NANO);
+    if (monitor == null) {
+      monitor = this.initMonitor();
+    }
+    assert monitor != null;
+    return monitor.forceRefresh(writerImportant, timeoutMs);
+  }
+
+  @Override
+  public void releaseResources() {
+    // do nothing
```

Comment:
We can release resources that belongs to this particular instance of MonitoringRdsHostListProvider. As I can see there's nothing that requires a special attention. Static resources, like monitoring threads and topology cache, can't be cleaned up and released since they might be in use by other connections.